org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/javasupport/JdtToBeBuiltComputer.java

```diff
@@ -193,4 +214,55 @@ public boolean isRejected(IFolder folder) {
 		boolean result = jdtHelper.isFromOutputPath(folder);
 		return result;
 	}
+	
+	/**
+	 * @since 2.19
+	 */
+	@Override
+	public void addInterestingProjects(IProject thisProject, Set<IProject> result) {
+		IJavaProject javaProject = JavaCore.create(thisProject);
+		if (javaProject instanceof JavaProject) {
+			Set<IProject> requiredProjects = getRequiredProjects((JavaProject) javaProject, thisProject.getWorkspace().getRoot());
+			projectDependencyGraph.putDependency(thisProject, requiredProjects);
+			result.addAll(requiredProjects);
+		}
+	}
+	
+	// JavaBuilder#build
+	private Set<IProject> getRequiredProjects(JavaProject javaProject, IWorkspaceRoot workspaceRoot) {
+		LinkedHashSet<IProject> result = new LinkedHashSet<>();
+		ExternalFoldersManager externalFoldersManager = JavaModelManager.getExternalManager();
+		try {
+			IClasspathEntry[] entries = javaProject.getExpandedClasspath();
+			for (int i = 0, l = entries.length; i < l; i++) {
+				IClasspathEntry entry = entries[i];
+				IPath path = entry.getPath();
+				IProject p = null;
+				switch (entry.getEntryKind()) {
+					case IClasspathEntry.CPE_PROJECT :
+						p = workspaceRoot.getProject(path.lastSegment()); // missing projects are considered too
+						if (((ClasspathEntry) entry).isOptional() && !JavaProject.hasJavaNature(p)) // except if entry is optional
```

Comment:
maybe we need to calculate transitive hull / reexports here too

Reply:
The logic is the same as for the Java builder. Could be, that we need another participant for plugin project semantics, though.